## 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np

# 세무서 코드

## 2. Load Data 1

In [2]:
data1 = pd.read_excel('../data/_nts_data_tax_code.xls')
print(data1.shape)
data1.head()

(131, 8)


세무서명                             도로명 주소  우편번호          전화번호          팩스번호  \
0  서울청            서울특별시 종로구 종로5길 86 (수송동)  3151   02-397-2200   02-722-0528   
1   종로              서울특별시 종로구 삼일대로 30길 22  3133   02-760-9200   02-744-4939   
2   중부                       서울 중구 소공로 70  4535  02-2260-9200  02-2278-0582   
3  남대문  서울특별시 중구 삼일대로 340 (저동1가) 나라키움저동빌딩  4551  02-2260-0200   02-755-7114   
4   용산        서울특별시 용산구 서빙고로24길15 (한강로3가)  4388   02-748-8200   02-792-2619   

   세무서코드   계좌번호                                               관할구역  
0    100  11895                                           서울특별시 전체  
1    101  11976                                       종로구 전역(87개동)  
2    201  11989  광희동 1,2가,  남대문로 2가,  남산동 1,2,3가,  남학동,  명동 1,2...  
3    104  11785  소공동남창동(남대문시장)북창동 \n회현동1-3가봉래동1,2가남대문로1,3,4,5가\...  
4    106  11947                                             용산구 전체

## 3. Reshape Data 1
### 3-1. EDA 1
- 원하는 컬럼만 특정해서 눈으로 한 번 check

In [5]:
data1[['세무서명', '세무서코드', '도로명 주소', '관할구역']]

세무서명  세무서코드                                도로명 주소  \
0    서울청    100               서울특별시 종로구 종로5길 86 (수송동)   
1     종로    101                 서울특별시 종로구 삼일대로 30길 22   
2     중부    201                          서울 중구 소공로 70   
3    남대문    104     서울특별시 중구 삼일대로 340 (저동1가) 나라키움저동빌딩   
4     용산    106           서울특별시 용산구 서빙고로24길15 (한강로3가)   
5     성북    209                    서울 성북구 삼선교로 16길 13   
6    서대문    110               서울특별시 서대문구 세무서길11 (홍제동)   
7     마포    105                     서울특별시 마포구 독막로 234   
8    영등포    107           서울특별시 영등포구 선유동1로 38 (당산동3가)   
9     강서    109              서울특별시 강서구 마곡서1로 60 (마곡동)   
10    양천    117              서울특별시 양천구 목동동로 165 (신정동)   
11    구로    113            서울특별시 영등포구 경인로 778 (문래동1가)   
12    동작    108             서울특별시 영등포구 대방천로 259 (신길동)   
13    금천    119        서울특별시 금천구 시흥대로152길 11-21 (독산동)   
14    관악    145          서울특별시 관악구 문성로 187(신림동 438-2)   
15    강남    211               서울특별시 강남구 학동로 425 (청담동)   
16    삼성    120    서울특별시 강남구 테헤란로 114 (역삼동) 역삼빌딩 5,6층   
17    서초    214    서울특별시 강남구 테헤란로 114 (역삼동) 역삼빌딩 3,4층   
18    역삼    220    서울특별시 강남구 테헤란로 114 (역삼동) 역삼빌딩 7,8층   
19    반포    114               서울특별시 서초구 방배로 163 (방배동)   
20    성동    206              서울특별시 성동구 광나루로 297 (송정동)   
21   동대문    204           서울특별시 동대문구 약령시길 159 (청량리1동)   
22    도봉    210               서울특별시 강북구 도봉로 117 (미아동)   
23    강동    212  서울특별시 강동구 천호대로 1139 강동그린타워(길동 459-3)   
24    송파    215               서울특별시 송파구 강동대로 62 (풍납동)   
25    노원    217              서울특별시 도봉구 노해로69길 14 (창동)   
26    잠실    230  서울특별시 송파구 강동대로 62 (풍납동, 송파세무서 청사 2층)   
27   중랑     146          서울특별시 중랑구 망우로 176(상봉동 137-1)   
28    은평    147              서울특별시 은평구 서오릉로7(응암동84-5)   
29   중부청    200        경기도 수원시 장안구 경수대로 1110-17 (파장동)   
..   ...    ...                                   ...   
101  북대구    504                대구광역시 북구 원대로 118 (침산동)   
102   수성    516           대구광역시 수성구 달구벌대로 2362(수성동3가)   
103   경산    515                  경상북도 경산시 박물관로 3 (사동)   
104   경주    505                경상북도 경주시 원화로 335 (성동동)   
105   포항    506                       경북 포항시 중앙로 346    
106   영덕    507                경상북도 영덕군 영덕읍 영덕로 35-11   
107   안동    508                경상북도 안동시 경동로 955(임시청사)   
108   김천    510                경상북도 김천시 평화길 128 (평화동)   
109   구미    513               경상북도 구미시 수출대로 179 (공단동)   
110   상주    511           경상북도 상주시 경상대로 3173-11 (만산동)   
111   영주    512                 경상북도 영주시 중앙로 15 (가흥동)   
112  부산청    600                부산광역시 연제구 연제로 12 (연산동)   
113  중부산    602               부산광역시 중구 흑교로 64 (보수동1가)   
114  서부산    603         부산광역시 서구 대영로 10(서대신동2가 288-2)   
115  부산진    605           부산광역시 동구 진성로 23 (수정동 247-7)   
116   수영    617          부산광역시 수영구 남천동로 19번길 28 (남천동)   
117  북부산    606               부산광역시 사상구 학감대로 263(감전동)   
118   동래    607          부산광역시 연제구 거제천로269번길 16 (거제동)   
119   금정    621             부산광역시 금정구 중앙대로 1636 (부곡동)   
120  해운대    623      부산광역시 해운대구 달맞이길62번길 38 4,5F (중동)   
121   울산    610                 울산광역시 남구 갈밭로 49 (삼산동)   
122  동울산    620                 울산광역시 북구 사청2길 7 (화봉동)   
123   마산    608       경상남도 창원시 마산합포구 해안대로 51 유로스퀘어 3층   
124   창원    609       경상남도 창원시 의창구 중앙대로209번길 16 (용호동)   
125   김해    615                경상남도 김해시 호계로 440 (부원동)   
126   거창    611                  경상남도 거창군 거창읍 상동2길 14   
127   통영    612              경상남도 통영시 무전5길 20-9 (무전동)   
128   진주    613           경상남도 진주시 진주대로908번길 15 (칠암동)   
129   양산    624                  경상남도 양산시 물금읍 증산역로135   
130   제주    616              제주특별자치도 제주시 청사로 59 (도남동)   

                                                  관할구역  
0                                             서울특별시 전체  
1                                         종로구 전역(87개동)  
2    광희동 1,2가,  남대문로 2가,  남산동 1,2,3가,  남학동,  명동 1,2...  
3    소공동남창동(남대문시장)북창동 \n회현동1-3가봉래동1,2가남대문로1,3,4,5가\...  
4                                               용산구 전체  
5                                            서울특별시 성북구  
6                                                 서대문구  
7            

In [6]:
# data1.to_excel('./semuseocode.xlsx', encoding='utf-8')

### 3-2. Create Dictionary
- key : value
    - 세무서코드 : 세무서명
- obj : 지역 별로 지사 선택

In [7]:
keys = list(data1['세무서코드'])
print(len(keys))
keys[:10]

131


[100, 101, 201, 104, 106, 209, 110, 105, 107, 109]

In [8]:
values = list(data1['세무서명'])
print(len(values))
values[:10]

131


['서울청', '종로', '중부', '남대문', '용산', '성북', '서대문', '마포', '영등포', '강서']

In [9]:
nts_code_dict = dict(zip(keys, values))
print(nts_code_dict)

{100: '서울청', 101: '종로', 201: '중부', 104: '남대문', 106: '용산', 209: '성북', 110: '서대문', 105: '마포', 107: '영등포', 109: '강서', 117: '양천', 113: '구로', 108: '동작', 119: '금천', 145: '관악', 211: '강남', 120: '삼성', 214: '서초', 220: '역삼', 114: '반포', 206: '성동', 204: '동대문', 210: '도봉', 212: '강동', 215: '송파', 217: '노원', 230: '잠실', 146: '중랑 ', 147: '은평', 200: '중부청', 121: '인천', 122: '북인천', 137: '서인천', 131: '남인천', 130: '부천', 123: '안양', 138: '동안양', 134: '안산', 124: '수원', 135: '동수원', 125: '평택', 129: '성남', 126: '이천', 127: '의정부', 231: '포천', 132: '남양주', 128: '고양', 141: '파주', 140: '시흥', 142: '용인', 143: '화성', 144: '분당', 221: '춘천', 223: '홍천', 224: '원주', 225: '영월', 222: '삼척', 226: '강릉', 227: '속초', 233: '경기광주', 234: '김포', 232: '동고양', 235: '광명', 236: '기흥', 300: '대전청', 305: '대전', 314: '서대전', 301: '청주', 317: '동청주', 302: '영동', 303: '충주', 304: '제천', 307: '공주', 308: '논산', 313: '보령', 316: '서산', 310: '홍성', 311: '예산', 312: '천안', 319: '아산', 318: '북대전', 320: '세종', 400: '광주청', 408: '광주', 409: '북광주', 410: '서광주', 401: '군산', 402: '전주', 418: '북

### Q1. 편하게 listing 하는 법 

In [15]:
seoul = []
incheon_gyeonggi = []
gangwon = []

busan = []

gyeongnam = []
ulsan = []
daegu_gyeongbuk = []

gwangju_jeolla = []

chungcheong = []

jeju = []

### 3-3. Test 1
- 임의의 사업자 번호가 개발한 코드에 맞게 추출되는지 확인하는 작업

# 사업자 정보
- 근로복지공단에서 사업자 번호를 통해 크롤링한 데이터를 불러와서 활용

## 4. Load Data 2

In [ ]:
data2 = pd.read_excel('./')
print(data2.shape)
data2.head()

## 5. Reshape Data 2
### 5-1. Rename columns name

In [ ]:
data2.colums = ['사업장명', '업종', '업종코드', '고용 상시 인원수', '산재 상시 인원수', '우편번호', '사업장 주소', '지사', '사업자번호']

### 5-2. EDA 2
- dtypes
- describe
- value_counts()

In [ ]:
print(data2.dtypes)
data2.describe

In [ ]:
data2['업종'].value_counts()

In [ ]:
data2['산재 상시 인원수'].value_counts()

In [ ]:
data2['지사'].value_counts()

### 5-3. Change Data types
- float -> int
- astype('int')

In [ ]:
data2['업종코드'].astype('int')

In [ ]:
data2['업종코드'].describe()

### 5-4. check null data

In [ ]:
data2['업종코드'] == np.nan

In [ ]:
data2[data2['업종코드'] == np.nan]

In [ ]:
data2['업종코드'].value_counts()

In [ ]:
data3 = data2.fillna(0)
print(data3.shape)
data3.head()

In [ ]:
data3[data3['업종코드'] == 0]

- 위(```data2[data2['업종코드'] == np.nan]```)에서는 안 나왔는데 여기서는 왜 나오지?

In [ ]:
data3.dtypes

In [ ]:
data3['업종코드'] = data3['업종코드'].astype('int')

In [ ]:
print(data3.shape)
print(data3.dtypes)
data3.head()

In [ ]:
data3[['사업장명', '사업자번호', '지사', '고용 상시 인원수', '산재 상시 인원수']]

## 6. Mapping Values
- 세무서 코드 list(세무서코드, 지역)와 사업자 정보 리스트의 정보(사업자 번호, 지사)가 맞는지 확인

### 6-1. Wrangling Data
- data3의 사업자정보
    - 사업자번호 - 지사
    - output : business_number_list

In [ ]:
data4 = data3[['사업장명','사업자번호', '지사']]
data5 = data4['사업자번호']

In [ ]:
print(type(data5))
data5.head()

In [ ]:
business_number_list = [business_number for business_number in data5]
print(len(business_number_list))
business_number_list[:10]

In [ ]:
type(business_number_list)

### Q2. numpy.int를 list indexing 하는 방법은? 

In [28]:
test_number = data1['세무서코드']

In [31]:
print(type(test_number))
test_number[1][:2]

<class 'pandas.core.series.Series'>


IndexError: invalid index to scalar variable.

In [33]:
test_number

0      100
1      101
2      201
3      104
4      106
5      209
6      110
7      105
8      107
9      109
10     117
11     113
12     108
13     119
14     145
15     211
16     120
17     214
18     220
19     114
20     206
21     204
22     210
23     212
24     215
25     217
26     230
27     146
28     147
29     200
      ... 
101    504
102    516
103    515
104    505
105    506
106    507
107    508
108    510
109    513
110    511
111    512
112    600
113    602
114    603
115    605
116    617
117    606
118    607
119    621
120    623
121    610
122    620
123    608
124    609
125    615
126    611
127    612
128    613
129    624
130    616
Name: 세무서코드, Length: 131, dtype: int64

In [40]:
test_number_list = list(data1['세무서코드'])
print(type(test_number_list))
# test_number_list[1][:2]

<class 'list'>


In [41]:
test_number_list[1]

101

In [43]:
str(test_number_list[1])[:2]

'10'

In [34]:
test_number_list

[100,
 101,
 201,
 104,
 106,
 209,
 110,
 105,
 107,
 109,
 117,
 113,
 108,
 119,
 145,
 211,
 120,
 214,
 220,
 114,
 206,
 204,
 210,
 212,
 215,
 217,
 230,
 146,
 147,
 200,
 121,
 122,
 137,
 131,
 130,
 123,
 138,
 134,
 124,
 135,
 125,
 129,
 126,
 127,
 231,
 132,
 128,
 141,
 140,
 142,
 143,
 144,
 221,
 223,
 224,
 225,
 222,
 226,
 227,
 233,
 234,
 232,
 235,
 236,
 300,
 305,
 314,
 301,
 317,
 302,
 303,
 304,
 307,
 308,
 313,
 316,
 310,
 311,
 312,
 319,
 318,
 320,
 400,
 408,
 409,
 410,
 401,
 402,
 418,
 403,
 404,
 407,
 411,
 412,
 415,
 416,
 417,
 500,
 502,
 503,
 514,
 504,
 516,
 515,
 505,
 506,
 507,
 508,
 510,
 513,
 511,
 512,
 600,
 602,
 603,
 605,
 617,
 606,
 607,
 621,
 623,
 610,
 620,
 608,
 609,
 615,
 611,
 612,
 613,
 624,
 616]

In [ ]:
type(business_number_list[1])

In [ ]:
business_number_list[1]

In [ ]:
business_number_list[:3]

In [ ]:
for i in range(len(business_number_list)):
    print(business_number_list[i])

In [ ]:
for i in range(len(business_number_list)):
    print(business_number_list[i][:3])

In [ ]:
[business_number_list[i][:3].sum for i in range(len(business_number_list))]

In [13]:
result = []
for i, v in enumerate(business_number_list):
#     print(i, v)
#     result.append(v[:3])
    print(business_number_list[i][:3])

### 6-2. 대안(이지만 문제가 존재)
- 147개의 사업자 번호를 하나하나 리스트에 넣었는데(copy&paste) 자동으로 할 수 있는 방법은?
    - 데이터가 무한대로 늘어날 경우가 분명 존재할 것이기에

In [ ]:
busi_num_list = []

In [ ]:
type(busi_num_list[0])

In [ ]:
busi_num_to_str_list = [str(i) for i in busi_num_list]

In [ ]:
print(busi_num_to_str_list[0])
print(busi_num_to_str_list[0][:3])
print(type(busi_num_to_str_list[0]))

### 6-3. Test3
- 1) data4 : 사업자 정보(사업장명, 사업자번호, 지사)
- 2) test_list : 로직으로 돌려서 나온 결과
- 3) result : 두 개를 합친 최종 결과물

In [ ]:
print(data4.shape)
data4.head()

### Q3. 편하게 count를 하는 방법은?
- 아래의 결과를 편하게 dataframe 형태로 만들 수 있는지?

In [ ]:
data4['지사'].unique()

In [ ]:
data4['지사'].value_counts()

In [ ]:
# a = []
# for k, v in enumerate(range(1, 10)):
#     a.append([k, v])
# a

In [ ]:
# pd.DataFrame(a)

In [ ]:
seoul = []
incheon_gyeonggi = []
gangwon = []

busan = []

gyeongnam = []
ulsan = []
daegu_gyeongbuk = []

gwangju_jeolla = []

chungcheong = []

jeju = []

In [17]:
busi_sample_list =[1010167107,
1010193167,
1010338381,
1010356040,
1010456450,
1010464377,
1010561926,
1010583672,
1010587741,
1010705383,
1010791647,
1010826236,
1010847129,
1010875227,
1011078101,
1011108903,
1011153937,
1011158121,
1011174359,
1031600289,
1040558546,
1040634901,
1041569474,
1050696370,
1050912653,
1050936320,
1050992467,
1051028506,
1051042584,
1051173328,
1051227419,
1051298928,
1051453279,
1051941609,
1052040116,
1052040682,
1052125581,
1052180643,
1052223824,
1053503584,
1053962989,
1054383318,
1057050128]

In [26]:
for i in range(len(busi_sample_list)):
    print(busi_sample_list[i][:3])

TypeError: 'int' object is not subscriptable

In [18]:
test_result = []

# for wrk in busi_num_to_str_list:
for wrk in busi_sample_list:
    if jeju[0] == wrk[:3]:
        test_result.append(['제주', wrk])
        
    for s in seoul:
        if s == wrk[:3]:
            test_result.append('서울', wrk)
            
    for ig in incheon_gyeonggi:
        if ig == wrk[:3]:
            test_result.append('인천경기', wrk)
            
    for g in gangwon:
        if g == wrk[:3]:
            test_result.append('강원', wrk)
            
    for b in busan:
        if b == wrk[:3]:
            test_result.append('부산', wrk)
            
    for g in gyeongnam :
        if g == wrk[:3]:
            test_result.append('경남', wrk)
            
    for u in ulsan:
        if u == wrk[:3]:
            test_result.append('울산', wrk)
            
    for dg in daegu_gyeongbuk:
        if dg == wrk[:3]:
            test_result.append('대구경북', wrk)
            
            
    for gj in gwangju_jeolla :
        if gj == wrk[:3]:
            test_result.append('광주전라', wrk)
            
    for c in chungcheong:
        if c == wrk[:3]:
            test_result.append('충청', wrk)
test_result

IndexError: list index out of range

In [ ]:
data6 = pd.DataFrame(test_result, columns=['지역', '사업자번호'])

In [ ]:
data6['지역'].value_counts()

In [ ]:
data4['지사'].value_counts()

In [ ]:
data4[data4['지사'] == '부산지역본부']

## 사업자번호 리스트, 크롤링 코드